wallet/src/actors/app/methods.rs

```diff
@@ -460,59 +463,35 @@ impl App {
         Box::new(f)
     }
 
-    /// Handle notifications received from the node.
-    pub fn handle_block_notification(&mut self, value: types::Json) -> Result<()> {
+    /// Handle new block notifications received from the node.
+    pub fn handle_block_notification(&self, value: types::Json) -> Result<()> {
         log::trace!("received block notification");
         let block = serde_json::from_value::<types::ChainBlock>(value).map_err(node_error)?;
-        // NOTE: Possible enhancement.
-        // Maybe is a good idea to use a shared reference Arc instead of cloning this vector of txns
-        // if this vector results to be too big, problem is that doing so conflicts with the internal
-        // Cell of the txns type which cannot be shared between threads.
-        let block_epoch = block.block_header.beacon.checkpoint;
-        // NOTE: We are calculating the hash of the block here (it's just the hash of the header)
-        // since it's not memoized and not doing it would imply that all threads indexing wallet
-        // UTXOs call this method over and over. If calculating the hash here degrades performance
-        // too much we should consider to calculate it in a worker thread instead and then proceed
-        // with indexing transactions.
-        let block_hash = block.hash().as_ref().to_vec();
-
-        // Block transactions to be indexed.
-        // NOTE: only `VttTransaction` and `DRTransaction` are currently supported.
-        let dr_txns = block
-            .txns
-            .data_request_txns
-            .into_iter()
-            .map(types::Transaction::from);
-        let vtt_txns = block
-            .txns
-            .value_transfer_txns
-            .into_iter()
-            .map(types::Transaction::from);
-        let block_txns = dr_txns.chain(vtt_txns).collect::<Vec<types::Transaction>>();
-
-        // Index block transactions on each wallet
-        for (id, wallet) in self.state.wallets() {
-            self.params.worker.do_send(worker::IndexTxns(
-                id.to_owned(),
-                wallet.clone(),
-                block_txns.clone(),
-                model::BlockInfo {
-                    epoch: block_epoch,
-                    hash: block_hash.clone(),
-                },
-            ));
-        }
 
-        log::trace!("notifying balances to sessions");
-        for (wallet, sink) in self.state.notifiable_wallets() {
-            self.params
-                .worker
-                .do_send(worker::NotifyBalance(wallet, sink));
+        for (_id, wallet) in self
```

Comment:
So we need to handle the block once for every wallet. What should happen if we create a new wallet after the initial block sync? Lose all the "old" UTXOs or sync again?

Reply:
Every new wallet is initiated in such a way that its `last_sync` is the genesis block. As a consequence, regardless of when the wallet is created or what blocks we have processed in the past, it syncs from the beginning and the final state is just right :)

Reply:
Needless to say that the fact that one wallet syncs at some point doesn't affect what was already processed for other wallets.

Reply:
Oh, nice! Do we want to store all the blocks somewhere to avoid asking the node more than once, or do we prefer to not store another copy of the blockchain in the wallet?

Reply:
Yes, let's avoid having the chain. At the end of the day that's the whole point of it being a separate component than the node — it should save users the disk space burden of running a full node.

On a side note, asking for the whole chain may sound overkill, but take into account that creating new wallets is done sparingly. This argument will specially get stronger if at some point Sheikah enables accounting, i.e. having more than one wallet for different purposes (daily expenses, hodling, etc.) derived from a single master key. In theory it's already supported by the wallet!